### PR TITLE
awscc - update action - ensure patch only against updatable properties

### LINKED ISSUE #9296: Set Athena Workgroup Encryption also tries to change readonly field
### Describe the bug

Hi, we are currently trying to set the encryption for our primary athena workgroups that are unencrypted. The policy looks like this:

```yaml
- name: set-athena-workgroup-encryption
  resource: awscc.athena_workgroup
  filters:
    - type: value
      key: Name
      value: "primary"
    - type: value
      key:  "WorkGroupConfiguration.ResultConfiguration.EncryptionConfiguration.EncryptionOption"
      value: absent
  actions:
    - type: update
      WorkGroupConfiguration:
        EnforceWorkGroupConfiguration: true
    - type: update
      WorkGroupConfiguration:
        ResultConfiguration:
          EncryptionConfiguration:
            EncryptionOption: SSE_S3
```

When executing this policy we get this error though: 

```
2024-02-15 09:59:25,626: custodian.commands:ERROR Error while executing policy set-athena-workgroup-encryption, continuing
Traceback (most recent call last):
  File "/home/.../policies/.venv/lib/python3.11/site-packages/c7n/commands.py", line 307, in run
    policy()
  File "/home/.../policies/.venv/lib/python3.11/site-packages/c7n/policy.py", line 1357, in __call__
    resources = mode.run()
                ^^^^^^^^^^
  File "/home/.../policies/.venv/lib/python3.11/site-packages/c7n/policy.py", line 364, in run
    results = a.process(resources)
              ^^^^^^^^^^^^^^^^^^^^
  File "/home/.../policies/.venv/lib/python3.11/site-packages/c7n_awscc/actions.py", line 43, in process
    client.update_resource(
  File "/home/.../policies/.venv/lib/python3.11/site-packages/botocore/client.py", line 553, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.../policies/.venv/lib/python3.11/site-packages/botocore/client.py", line 1009, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (ValidationException) when calling the UpdateResource operation: Invalid patch update: readOnlyProperties [/properties/WorkGroupConfiguration/EngineVersion/EffectiveEngineVersion] cannot be updated
2024-02-15 09:59:25,627: custodian.commands:ERROR The following policies had errors while executing
 - set-athena-workgroup-encryption
```

But we don't want to update the engine version itself.

### What did you expect to happen?

We expected the policy to update the encryption setting and not touch the engine version, because the attribute was not specified in our policy

### Cloud Provider

Amazon Web Services (AWS)

### Cloud Custodian version and dependency information

```shell
Custodian:   0.9.34
Python:      3.11.4 (main, Dec  7 2023, 15:43:41) [GCC 12.3.0]
Platform:    posix.uname_result(sysname='Linux', nodename='marcel', release='6.2.0-39-generic', version='#40-Ubuntu SMP PREEMPT_DYNAMIC Tue Nov 14 14:18:00 UTC 2023', machine='x86_64')
Using venv:  True
Docker: False
Installed:

argcomplete==3.2.1
attrs==23.2.0
boto3==1.34.21
botocore==1.34.21
docutils==0.18.1
importlib-metadata==6.11.0
jmespath==1.0.1
jsonschema==4.21.0
jsonschema-specifications==2023.12.1
python-dateutil==2.8.2
pyyaml==6.0.1
referencing==0.31.1
rpds-py==0.17.1
s3transfer==0.10.0
six==1.16.0
tabulate==0.9.0
urllib3==1.26.18
zipp==3.17.0
```

### Policy

```shell
- name: set-athena-workgroup-encryption
  resource: awscc.athena_workgroup
  filters:
    - type: value
      key: Name
      value: "primary"
    - type: value
      key:  "WorkGroupConfiguration.ResultConfiguration.EncryptionConfiguration.EncryptionOption"
      value: absent
  actions:
    - type: update
      WorkGroupConfiguration:
        EnforceWorkGroupConfiguration: true
    - type: update
      WorkGroupConfiguration:
        ResultConfiguration:
          EncryptionConfiguration:
            EncryptionOption: SSE_S3
```

### Relevant log/traceback output

```shell
2024-02-15 09:59:25,626: custodian.commands:ERROR Error while executing policy set-athena-workgroup-encryption, continuing
Traceback (most recent call last):
  File "/home/.../policies/.venv/lib/python3.11/site-packages/c7n/commands.py", line 307, in run
    policy()
  File "/home/.../policies/.venv/lib/python3.11/site-packages/c7n/policy.py", line 1357, in __call__
    resources = mode.run()
                ^^^^^^^^^^
  File "/home/.../policies/.venv/lib/python3.11/site-packages/c7n/policy.py", line 364, in run
    results = a.process(resources)
              ^^^^^^^^^^^^^^^^^^^^
  File "/home/.../policies/.venv/lib/python3.11/site-packages/c7n_awscc/actions.py", line 43, in process
    client.update_resource(
  File "/home/.../policies/.venv/lib/python3.11/site-packages/botocore/client.py", line 553, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/.../policies/.venv/lib/python3.11/site-packages/botocore/client.py", line 1009, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (ValidationException) when calling the UpdateResource operation: Invalid patch update: readOnlyProperties [/properties/WorkGroupConfiguration/EngineVersion/EffectiveEngineVersion] cannot be updated
2024-02-15 09:59:25,627: custodian.commands:ERROR The following policies had errors while executing
 - set-athena-workgroup-encryption
```

### Extra information or context

We tried to use the update attributes like this 

```yaml
    - type: update
      WorkGroupConfigurationUpdates:
        EnforceWorkGroupConfiguration: true
        ResultConfigurationUpdates:
          EncryptionConfiguration:
            EncryptionOption: SSE_S3
```

but there is currently a bug in AWS which resets the workgroup right after the operation. We are in communication with AWS Support there, but in the meantime we tried to make it work with the approach described above.

### PR DESCRIPTION
**Pull request by kapilt**
closes #9296

### CODE REVIEW

File: tools/c7n_awscc/c7n_awscc/actions.py
Comment (ajkerrigan) [inline tools/c7n_awscc/c7n_awscc/actions.py]: If we're only excluding or override top-level keys, could we still get the key skipping benefits without deepcopy? Something like:

```suggestion
        # the action's schema reflects updatable properties
        updatable = {k for k in self.schema["properties"].items() if k != "type"}
        current = {k: r[k] for k in updatable}
```

**Review by ajkerrigan — APPROVED** [review]
Just a question inline around the need for deepcopy, otherwise lgtm once checks are passing :+1:

File: tools/c7n_awscc/c7n_awscc/actions.py
Comment (kapilt) [reply tools/c7n_awscc/c7n_awscc/actions.py]: yeah.. we might be able to get away with the need for deep copy, since we're only doing top level key replacement here, my concern is if we need to be doing partial updates, see referenced issue `WorkGroupConfiguration` then we may need to be doing patch updates into dicts in which case we do need a deep copy to preserve the original.

Comment (ajkerrigan) [reply tools/c7n_awscc/c7n_awscc/actions.py]: Ahhh gotcha, okie safety wins. Thanks for the clarification :+1:

Comment (kapilt) [reply tools/c7n_awscc/c7n_awscc/actions.py]: went ahead and added patch functionality, so it might be worth revisiting this as we're only doing top level key/value replacement, re can get away with shallow copy.

**Review by ajkerrigan — APPROVED** [review]
This looks good to me, thanks! :tada: 

Two things I was noodling on were:

- Would this make more sense as a separate `patch` action?
  - Probably not
- Can we make this a bit more streamlined for policy authors rather than exposing full/explicit jsonpatch syntax?
  - Maaaybe, but the cost would be making things less obvious or too magical, particularly for the `op: remove` case

**Comment by kapilt** [comment]
@ajkerrigan I think you answered your own questions ;-) but let me know if you want some additional context. fwiw, we might need to extend this action in the future to take into consideration in flight operations to the resources.

**Pull request merged**
